src/steamship/agents/context/context.py

```diff
@@ -0,0 +1,100 @@
+from abc import ABC
+from typing import Any, Callable, Dict, List, Optional, Tuple
+
+from pydantic.main import BaseModel
+
+from steamship import Block, PluginInstance, Steamship, Tag, Task
+from steamship.agents.context.chathistory import ChatHistory
+
+
+class BaseTool(BaseModel, ABC):
+    pass
+
+
+class ToolBinding(BaseModel):
+    tool: BaseTool
+    inputs: List[Block]
+
+
+ToolInputs = List[Block]
+ToolOutputs = List[Block]
+AgentStep = Tuple[ToolBinding, ToolOutputs]
+AgentSteps = List[AgentStep]
+Metadata = Dict[str, Any]
+EmitFunc = Callable[[ToolOutputs, Metadata], None]
+
+
+class AgentContext:
+    # AgentContext is passed as a parameter for any `run_agent()` call on the `AgentService`
+    # as well as any `Tool.run()` call. It provides access to arbitrary metadata about the
+    # context of any agent invocation, as well as anything relevant to the `AgentService`
+    # instance etc.
+    #
+    # It also provides access to two-distinct types of "history" or "memory":
+    # (1) user <-> package interactions
+    # (2) agent and tool interactions
+    #
+    # It is meant to be persisted (via a KeyValueStore or otherwise) to allow for
+    # restarts, interruptions, etc.
+
+    # A steamship-assigned ID for this context object
+    @property
+    def id(self) -> str:
+        return self.chat_history.file.id
+
+    metadata: Metadata = {}
+
+    client: Steamship
+
+    # User<->Package chat history (NOT Agent<-->Tool history)
+    # NB: This is distinct from any sort of history related to agent execution
+    # Primarily used to ferry prompts into agent executions AND
+    # to allow for some sort of referential lookup of context as input to the agent
+    chat_history: ChatHistory
+
+    # Provides the overall goal of the agent
+    @property
+    def initial_prompt(self) -> Optional[Block]:
+        return self.chat_history.initial_system_prompt
+
+    # instead of saving the execution history via File as text, this chooses to represent
```

Comment:
I know we've got a TON of different ideas on how to best accomplish this, so rather than jump in advocating for any particular one of them... I wanted to jump in and say:

- I think we're within the threshold of: "Any we've been discussing sound fine, and we can change it later -- even if it breaks backward compatibility." 

Feels like one of those totally OK "cost of iteration / growing" things.